File: backend/app/models/graph.py
"""Graph-related models."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class GraphInfo(BaseModel):
    """Graph metadata."""

    name: str
    id: Optional[str] = None


class NodeLabel(BaseModel):
    """Node label information."""

    label: str
    count: int
    properties: List[str] = Field(default_factory=list)


class PropertyStatistics(BaseModel):
    """Statistics for a numeric property."""

    property: str
    min: Optional[float] = None
    max: Optional[float] = None


class EdgeLabel(BaseModel):
    """Edge label information."""

    label: str
    count: int
    properties: List[str] = Field(default_factory=list)
    property_statistics: List[PropertyStatistics] = Field(
        default_factory=list, description="Statistics for numeric properties (min/max)"
    )


class GraphMetadata(BaseModel):
    """Complete graph metadata."""

    graph_name: str
    node_labels: List[NodeLabel]
    edge_labels: List[EdgeLabel]
    role: Optional[str] = None


class MetaGraphEdge(BaseModel):
    """Meta-graph relationship pattern."""

    source_label: str
    target_label: str
    edge_label: str
    count: int


class MetaGraphResponse(BaseModel):
    """Meta-graph view of label relationships."""

    graph_name: str
    relationships: List[MetaGraphEdge]


class NodeExpandRequest(BaseModel):
    """Request to expand neighborhood of a node."""

    depth: int = Field(default=1, ge=1, le=5, description="Expansion depth (1-5)")
    limit: int = Field(default=100, ge=1, le=1000, description="Maximum nodes to return (1-1000)")


class NodeExpandResponse(BaseModel):
    """Response from node expansion."""

    nodes: List[Dict[str, Any]] = Field(..., description="Expanded nodes")
    edges: List[Dict[str, Any]] = Field(..., description="Expanded edges")
    node_count: int = Field(..., description="Number of nodes returned")
    edge_count: int = Field(..., description="Number of edges returned")


class NodeDeleteRequest(BaseModel):
    """Request to delete a node."""

    detach: bool = Field(
        default=False, description="If true, delete node and all its relationships. If false, only delete if no relationships exist."
    )


class NodeDeleteResponse(BaseModel):
    """Response from node deletion."""

    deleted: bool = Field(..., description="Whether the node was deleted")
    node_id: str = Field(..., description="ID of the deleted node")
    edges_deleted: int = Field(default=0, description="Number of edges deleted (if detach=true)")


class ShortestPathRequest(BaseModel):
    """Request to find shortest path between two nodes."""

    source_id: int = Field(..., description="Source node ID")
    target_id: int = Field(..., description="Target node ID")
    max_depth: int = Field(
        default=10,
        ge=1,
        le=20,
        description="Maximum path length in hops (1-20)",
    )


class ShortestPathResponse(BaseModel):
    """Response from shortest path query."""

    path: Optional[List[Dict[str, Any]]] = Field(
        default=None,
        description="Nodes and edges in path order, or None if no path found",
    )
    path_length: int = Field(
        default=0,
        description="Number of edges in the path (0 if no path)",
    )
    nodes: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Nodes in path order",
    )
    edges: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Edges in path order",
    )

File: frontend/src/services/graph.ts
/**
 * Graph metadata API.
 */

import apiClient from './api'

export interface GraphInfo {
  name: string
  id?: string
}

export interface NodeLabel {
  label: string
  count: number
  properties: string[]
}

export interface PropertyStatistics {
  property: string
  min: number | null
  max: number | null
}

export interface EdgeLabel {
  label: string
  count: number
  properties: string[]
  property_statistics?: PropertyStatistics[]
}

export interface GraphMetadata {
  graph_name: string
  node_labels: NodeLabel[]
  edge_labels: EdgeLabel[]
  role?: string
}

export interface MetaGraphEdge {
  source_label: string
  target_label: string
  edge_label: string
  count: number
}

export interface MetaGraphResponse {
  graph_name: string
  relationships: MetaGraphEdge[]
}

export interface NodeExpandRequest {
  depth?: number
  limit?: number
}

export interface GraphNode {
  id: string
  label: string
  properties: Record<string, unknown>
  type: string
}

export interface GraphEdge {
  id: string
  label: string
  source: string
  target: string
  properties: Record<string, unknown>
  type: string
}

export interface NodeExpandResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
  node_count: number
  edge_count: number
}

export interface NodeDeleteRequest {
  detach?: boolean
}

export interface NodeDeleteResponse {
  deleted: boolean
  node_id: string
  edges_deleted: number
}

export const graphAPI = {
  listGraphs: async (): Promise<GraphInfo[]> => {
    const response = await apiClient.get<GraphInfo[]>('/graphs')
    return response.data
  },

  getMetadata: async (graphName: string): Promise<GraphMetadata> => {
    const response = await apiClient.get<GraphMetadata>(
      `/graphs/${encodeURIComponent(graphName)}/metadata`
    )
    return response.data
  },

  getMetaGraph: async (graphName: string): Promise<MetaGraphResponse> => {
    const response = await apiClient.get<MetaGraphResponse>(
      `/graphs/${encodeURIComponent(graphName)}/meta-graph`
    )
    return response.data
  },

  expandNode: async (
    graphName: string,
    nodeId: string,
    request: NodeExpandRequest = {}
  ): Promise<NodeExpandResponse> => {
    const response = await apiClient.post<NodeExpandResponse>(
      `/graphs/${encodeURIComponent(graphName)}/nodes/${encodeURIComponent(nodeId)}/expand`,
      request
    )
    return response.data
  },

  deleteNode: async (
    graphName: string,
    nodeId: string,
    request: NodeDeleteRequest = {}
  ): Promise<NodeDeleteResponse> => {
    // Use query params for the detach option (FastAPI DELETE endpoints accept query params)
    const params = new URLSearchParams()
    if (request.detach !== undefined) {
      params.append('detach', String(request.detach))
    }
    const url = `/graphs/${encodeURIComponent(graphName)}/nodes/${encodeURIComponent(nodeId)}`
    const response = await apiClient.delete<NodeDeleteResponse>(
      params.toString() ? `${url}?${params.toString()}` : url
    )
    return response.data
  },
}

